# Pandas

Họ và tên: Ngô Thị Minh Thư - PTDL K55

Email: ngothiminhthu02@gmail.com

## 1. Giới thiệu Pandas
Pandas là gì? 

Pandas là một thư viện Python cung cấp cấu trúc dữ liệu nhanh, mạnh mẽ, linh hoạt và mang ý nghĩa. Pandas được thiết kế để làm việc dễ dàng và trực tiếp với dữ liệu có cấu trúc (dạng bảng, đa chiều) và dữ liệu chuỗi thời gian.

Pands là thư viện mã nguồn mở (open source), xử lý linh hoạt dữ liệu có cấu trúc, Trong Data Science pandas được sử dụng như là một thư viện chính để nhập xuất dữ liệu, tiền xử lý dữ liệu thô

Use the following import convention:

In [1]:
import numpy as np
import pandas as pd

## 2. Tạo đối tượng cơ bản trong Pandas
### 2.1 Series
Một Series trong pandas rất giống với một mảng NumPy (trên thực tế, nó được xây dựng trên đối tượng mảng NumPy). Điều khác biệt giữa mảng NumPy với một series là series có thể có các nhãn trục, nghĩa là nó có thể được lập chỉ mục bởi một nhãn, thay vì chỉ một vị trí số. Nó cũng không cần phải chứa dữ liệu số, nó có thể chứa bất kỳ Đối tượng Python nào tùy ý.

Bạn có thể tạo một series từ list, mảng numpy hoặc dict

In [3]:
labels = ['a','b','c']
my_list = [10,20,30]
arr = np.array([10,20,30])
dict = {'a':10,'b':20,'c':30}

Sử dụng list

In [5]:
pd.Series(data=my_list)

0    10
1    20
2    30
dtype: int64

Có gắn nhãn cho series

In [6]:
pd.Series(data=my_list,index=labels)

a    10
b    20
c    30
dtype: int64

Có thể viết gọn không cần truyền tham số mà chỉ cần đối số

In [7]:
pd.Series(my_list,labels)

a    10
b    20
c    30
dtype: int64

Tạo series từ một mảng numpy ( array numpy)

In [8]:
pd.Series(arr,labels)

a    10
b    20
c    30
dtype: int32

Tạo từ 1 dictionary

In [9]:
pd.Series(dict)

a    10
b    20
c    30
dtype: int64

#### Kiểu dữ liệu trong Series
Pandas có thể chứa nhiều kiểu dữ liệu khác nhau như số (integer), chuỗi (string), kiểu boolean (True, false), kiểu dữ liệu NaN.

In [10]:
pd.Series(data=labels)

0    a
1    b
2    c
dtype: object

### 2.2 Dataframe
Data Frame là gì?

DataFrame là một loạt các Series được ghép lại với nhau để chia sẻ cùng một chỉ mục (index), nó tương tự như trên bảng tính excel mà ta đã từng gặp qua. Bao gồm các dòng và cột, các cột ở đây tương ứng như một series mà ta đã học trong bài trước.

<img src="files/image.png">
Tạo 1 Data Frame từ list

In [14]:
list = [[1,6],[2,7],[3,8],[4,9],[5,10]]
dict = {'A': [1,2,3,4,5], "B": [6,7,8,9,10]}
column = ['A','B']
index = ['I1', 'I2',"I3",'I4',"I5"]

In [15]:
df = pd.DataFrame(list)
print(df)

   0   1
0  1   6
1  2   7
2  3   8
3  4   9
4  5  10


Mặc định khi tạo Data Frame sẽ gán giá trị theo thứ tự từ 0 trở đi cho cột và dòng, nếu muốn thay đổi tên cột và dòng ta làm như sau

In [16]:
df = pd.DataFrame(dict, index = index, columns= column)
df

A   B
I1  1   6
I2  2   7
I3  3   8
I4  4   9
I5  5  10

Tạo 1 Data Frame từ Dictionary

In [17]:
df = pd.DataFrame(dict)
print(df)

   A   B
0  1   6
1  2   7
2  3   8
3  4   9
4  5  10


Lúc này, các khóa trong dict mặc định là tên cột, index được gán mặc định là thứ tự từ 0 trở đi

Ta có thể gán index theo cú pháp bên dưới

In [18]:
df = pd.DataFrame(dict, index = index)
print(df)

    A   B
I1  1   6
I2  2   7
I3  3   8
I4  4   9
I5  5  10


Phương thức khác để tạo một data frame từ numpy

In [19]:
from numpy.random import randn # Lấy thư viện random
np.random.seed(101) # cố định số ngẫu nhiên tạo ra từ random
df = pd.DataFrame(randn(5,4),index='A B C D E'.split(),columns='W X Y Z'.split())
print(df)

          W         X         Y         Z
A  2.706850  0.628133  0.907969  0.503826
B  0.651118 -0.319318 -0.848077  0.605965
C -2.018168  0.740122  0.528813 -0.589001
D  0.188695 -0.758872 -0.933237  0.955057
E  0.190794  1.978757  2.605967  0.683509


### 3. Đọc dữ liệu từ các nguồn khác nhau
#### Đọc dữ liệu từ file CSV
With CSV files all you need is a single line to load in the data:

In [2]:
df = pd.read_csv('purchases.csv')

df

Unnamed: 0  apples  oranges
0       June       3        0
1     Robert       2        3
2       Lily       0        7
3      David       1        2

CSVs don't have indexes like our DataFrames, so all we need to do is just designate the index_col when reading:
Here we're setting the index to be column zero.

In [3]:
df = pd.read_csv('purchases.csv', index_col=0)
df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

#### Đọc dữ liệu từ file JSON
If you have a JSON file — which is essentially a stored Python dict — pandas can read this just as easily:

In [4]:
df = pd.read_json('purchases.json')
df

apples  oranges
June         3        0
Robert       2        3
Lily         0        7
David        1        2

Pandas will try to figure out how to create a DataFrame by analyzing structure of your JSON, and sometimes it doesn't get it right. Often you'll need to set the orient keyword argument depending on the structure, so check out read_json docs about that argument to see which orientation you're using.

#### Đọc dữ liệu từ SQL database
If you’re working with data from a SQL database you need to first establish a connection using an appropriate Python library, then pass a query to pandas. Here we'll use SQLite to demonstrate.

First, we need pysqlite3 installed, so run this command in your terminal:

In [22]:
!pip install pysqlite3

  Using cached pysqlite3-0.4.7.tar.gz (40 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Running setup.py clean for pysqlite3
Failed to build pysqlite3
  Running setup.py install for pysqlite3: started
  Running setup.py install for pysqlite3: finished with status 'done'


  Command arguments: 'C:\Users\ADMIN\anaconda3\python.exe' -u -c '
  exec(compile('"'"''"'"''"'"'
  # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
  #
  # - It imports setuptools before invoking setup.py, to enable projects that directly
  #   import from `distutils.core` to work with newer packaging standards.
  # - It provides a clear error message when setuptools is not installed.
  # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
  #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
  #     manifest_maker: standard file '"'"'-c'"'"' not found".
  # - It generates a shim setup.py, for handling setup.cfg-only projects.
  import os, sys, tokenize
  
  try:
      import setuptools
  except ImportError as error:
      print(
          "ERROR: Can not execute `setup.py` since setuptools is not available in "
          "the build environment.",
          file=sys.stderr,
      )
      sys.e

sqlite3 is used to create a connection to a database which we can then use to generate a DataFrame through a SELECT query.

So first we'll make a connection to a SQLite database file:

In [5]:
import sqlite3

con = sqlite3.connect("database.db")

If you have data in PostgreSQL, MySQL, or some other SQL server, you'll need to obtain the right Python library to make a connection. For example, psycopg2 (link) is a commonly used library for making connections to PostgreSQL. Furthermore, you would make a connection to a database URI instead of a file.

In this SQLite database we have a table called purchases, and our index is in a column called "index".

By passing a SELECT query and our con, we can read from the purchases table:

In [6]:
df = pd.read_sql_query("SELECT * FROM purchases", con)

df

index  apples  oranges
0    June       3        0
1  Robert       2        3
2    Lily       0        7
3   David       1        2

Just like with CSVs, we could pass index_col='index', but we can also set an index after-the-fact:

In [7]:
df = df.set_index('index')

df

apples  oranges
index                  
June         3        0
Robert       2        3
Lily         0        7
David        1        2

Converting back to a CSV, JSON, or SQL

So after extensive work on cleaning your data, you’re now ready to save it as a file of your choice. Similar to the ways we read in data, pandas provides intuitive commands to save it.

When we save JSON and CSV files, all we have to input into those functions is our desired filename with the appropriate file extension. With SQL, we’re not creating a new file but instead inserting a new table into the database using our con variable from before.

In [8]:
df.to_csv('new_purchases.csv')

df.to_json('new_purchases.json')

df.to_sql('new_purchases', con)

4

### 4. Quan sát và truy xuất dữ liệu trong DataFrame
Let's load in the IMDB movies dataset to begin:

In [9]:
movies_df = pd.read_csv("IMDB-Movie-Data.csv", index_col="Title")

#### Viewing your data

The first thing to do when opening a new dataset is print out a few rows to keep as a visual reference. We accomplish this with .head():m

In [10]:
movies_df.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

.head() outputs the first five rows of your DataFrame by default, but we could also pass a number as well: movies_df.head(10) would output the top ten rows, for example.

To see the last five rows use .tail(). tail() also accepts a number, and in this case we printing the bottom two rows.:

In [11]:
movies_df.tail(2)

Rank                  Genre  \
Title                                       
Search Party   999       Adventure,Comedy   
Nine Lives    1000  Comedy,Family,Fantasy   

                                                    Description  \
Title                                                             
Search Party  A pair of friends embark on a mission to reuni...   
Nine Lives    A stuffy businessman finds himself trapped ins...   

                      Director  \
Title                            
Search Party    Scot Armstrong   
Nine Lives    Barry Sonnenfeld   

                                                         Actors  Year  \
Title                                                                   
Search Party  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
Nine Lives    Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

              Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Title                                                                          
Search Party                 93     5.6   4881                 NaN       22.0  
Nine Lives                   87     5.3  12435               19.64       11.0

#### Getting info about your data

.info() should be one of the very first commands you run after loading your data:

In [12]:
movies_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 93.8+ KB


.info() provides the essential details about your dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory your DataFrame is using.

Notice in our movies dataset we have some obvious missing values in the Revenue and Metascore columns. We'll look at how to handle those in a bit.

Seeing the datatype quickly is actually quite useful. Imagine you just imported some JSON and the integers were recorded as strings. You go to do some arithmetic and find an "unsupported operand" Exception because you can't do math with strings. Calling .info() will quickly point out that your column you thought was all integers are actually string objects.

Another fast and useful attribute is .shape, which outputs just a tuple of (rows, columns):

In [13]:
movies_df.shape

(1000, 11)

#### Truy xuất dữ liệu
Dựa vào tên column ta truy suất cột (lúc này trả về cho ta là 1 series)

In [14]:
movies_df['Genre']
# hoặc ta có thể truy suất theo thuộc tính cột 
movies_df.Genre # nếu tên cột không có khoảng cách

Title
Guardians of the Galaxy     Action,Adventure,Sci-Fi
Prometheus                 Adventure,Mystery,Sci-Fi
Split                               Horror,Thriller
Sing                        Animation,Comedy,Family
Suicide Squad              Action,Adventure,Fantasy
                                     ...           
Secret in Their Eyes            Crime,Drama,Mystery
Hostel: Part II                              Horror
Step Up 2: The Streets          Drama,Music,Romance
Search Party                       Adventure,Comedy
Nine Lives                    Comedy,Family,Fantasy
Name: Genre, Length: 1000, dtype: object

Ta có thể truy xuất nhiều cột

In [16]:
# bằng cách sử dụng 2 dấu ngoặc vuông
movies_df[['Genre','Director']]

Genre              Director
Title                                                                  
Guardians of the Galaxy   Action,Adventure,Sci-Fi            James Gunn
Prometheus               Adventure,Mystery,Sci-Fi          Ridley Scott
Split                             Horror,Thriller    M. Night Shyamalan
Sing                      Animation,Comedy,Family  Christophe Lourdelet
Suicide Squad            Action,Adventure,Fantasy            David Ayer
...                                           ...                   ...
Secret in Their Eyes          Crime,Drama,Mystery             Billy Ray
Hostel: Part II                            Horror              Eli Roth
Step Up 2: The Streets        Drama,Music,Romance            Jon M. Chu
Search Party                     Adventure,Comedy        Scot Armstrong
Nine Lives                  Comedy,Family,Fantasy      Barry Sonnenfeld

[1000 rows x 2 columns]

Để truy xuất dữ liệu theo dòng, ta dùng hàm loc()

In [17]:
movies_df.loc[["Search Party","Nine Lives"]]

Rank                  Genre  \
Title                                       
Search Party   999       Adventure,Comedy   
Nine Lives    1000  Comedy,Family,Fantasy   

                                                    Description  \
Title                                                             
Search Party  A pair of friends embark on a mission to reuni...   
Nine Lives    A stuffy businessman finds himself trapped ins...   

                      Director  \
Title                            
Search Party    Scot Armstrong   
Nine Lives    Barry Sonnenfeld   

                                                         Actors  Year  \
Title                                                                   
Search Party  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
Nine Lives    Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

              Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Title                                                                          
Search Party                 93     5.6   4881                 NaN       22.0  
Nine Lives                   87     5.3  12435               19.64       11.0

Xóa cột trong Data frame

In [28]:
movies_df.drop('Metascore',axis=1,inplace=True)
print(movies_df)

                         Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...  

In [29]:
movies_df

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  Year  Runtime (Minutes)  \
Title                                                                    
Guardians of the Galaxy            James Gunn  2014                121   
Prometheus                       Ridley Scott  2012                124   
Split                      M. Night Shyamalan  2016                117   
Sing                     Christophe Lourdelet  2016                108   
Suicide Squad                      David Ayer  2016                123   
...                                       ...   ...                ...   
Secret in Their Eyes                Billy Ray  2015                111   
Hostel: Part II                      Eli Roth  2007                 94   
Step Up 2: The Streets             Jon M. Chu  2008                 98   
Search Party                   Scot Armstrong  2014                 93   
Nine Lives                   Barry Sonnenfeld  2016                 87   

                         Rating   Votes  Revenue (Millions)  \
Title                                                         
Guardians of the Galaxy     8.1  757074              333.13   
Prometheus                  7.0  485820              126.46   
Split                       7.3  157606              138.12   
Sing                        7.2   60545              270.32   
Suicide Squad               6.2  393727              325.02   
...                         ...     ...                 ...   
Secret in Their Eyes        6.2   27585                 NaN   
Hostel: Part II             5.5   73152               17.54   
Step Up 2: The Streets      6.2   70699               58.01   
Search Party                5.6    4881                 NaN   
Nine Lives                  5.3   12435               19.64   

                                                                       new  
Title                                                                       
Guardians of the Galaxy  Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus               Title\nGuardians of the Galaxy       1\nPromet...  
Split                    Title\nGuardians of the Galaxy       1\nPromet...  
Sing                     Title\nGuardians of the Galaxy       1\nPromet...

Xóa dòng 

In [30]:
movies_df.drop('Search Party',axis=0)
movies_df

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  Year  Runtime (Minutes)  \
Title                                                                    
Guardians of the Galaxy            James Gunn  2014                121   
Prometheus                       Ridley Scott  2012                124   
Split                      M. Night Shyamalan  2016                117   
Sing                     Christophe Lourdelet  2016                108   
Suicide Squad                      David Ayer  2016                123   
...                                       ...   ...                ...   
Secret in Their Eyes                Billy Ray  2015                111   
Hostel: Part II                      Eli Roth  2007                 94   
Step Up 2: The Streets             Jon M. Chu  2008                 98   
Search Party                   Scot Armstrong  2014                 93   
Nine Lives                   Barry Sonnenfeld  2016                 87   

                         Rating   Votes  Revenue (Millions)  \
Title                                                         
Guardians of the Galaxy     8.1  757074              333.13   
Prometheus                  7.0  485820              126.46   
Split                       7.3  157606              138.12   
Sing                        7.2   60545              270.32   
Suicide Squad               6.2  393727              325.02   
...                         ...     ...                 ...   
Secret in Their Eyes        6.2   27585                 NaN   
Hostel: Part II             5.5   73152               17.54   
Step Up 2: The Streets      6.2   70699               58.01   
Search Party                5.6    4881                 NaN   
Nine Lives                  5.3   12435               19.64   

                                                                       new  
Title                                                                       
Guardians of the Galaxy  Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus               Title\nGuardians of the Galaxy       1\nPromet...  
Split                    Title\nGuardians of the Galaxy       1\nPromet...  
Sing                     Title\nGuardians of the Galaxy       1\nPromet...

#### Truy xuất giá trị có điều kiện trong Data Frame

In [31]:
# Truy xuất giá trị Rank nhỏ hơn 500 trong data frame 
movies_df[movies_df['Rank']<500]

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Under the Skin            495       Drama,Horror,Sci-Fi   
I Am Legend               496       Drama,Horror,Sci-Fi   
Men in Black 3            497   Action,Adventure,Comedy   
Super 8                   498   Mystery,Sci-Fi,Thriller   
Law Abiding Citizen       499      Crime,Drama,Thriller   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Under the Skin           A mysterious young woman seduces lonely men in...   
I Am Legend              Years after a plague kills most of humanity an...   
Men in Black 3           Agent J travels in time to M.I.B.'s early days...   
Super 8                  During the summer of 1979, a group of friends ...   
Law Abiding Citizen      A frustrated man decides to take justice into ...   

                                     Director  Year  Runtime (Minutes)  \
Title                                                                    
Guardians of the Galaxy            James Gunn  2014                121   
Prometheus                       Ridley Scott  2012                124   
Split                      M. Night Shyamalan  2016                117   
Sing                     Christophe Lourdelet  2016                108   
Suicide Squad                      David Ayer  2016                123   
...                                       ...   ...                ...   
Under the Skin                Jonathan Glazer  2013                108   
I Am Legend                  Francis Lawrence  2007                101   
Men in Black 3               Barry Sonnenfeld  2012                106   
Super 8                           J.J. Abrams  2011                112   
Law Abiding Citizen              F. Gary Gray  2009                109   

                         Rating   Votes  Revenue (Millions)  \
Title                                                         
Guardians of the Galaxy     8.1  757074              333.13   
Prometheus                  7.0  485820              126.46   
Split                       7.3  157606              138.12   
Sing                        7.2   60545              270.32   
Suicide Squad               6.2  393727              325.02   
...                         ...     ...                 ...   
Under the Skin              6.3   94707                2.61   
I Am Legend                 7.2  565721              256.39   
Men in Black 3              6.8  278379              179.02   
Super 8                     7.1  298913              126.98   
Law Abiding Citizen         7.4  228339               73.34   

                                                                       new  
Title                                                                       
Guardians of the Galaxy  Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus               Title\nGuardians of the Galaxy       1\nPromet...  
Split                    Title\nGuardians of the Galaxy       1\nPromet...  
Sing                     Title\nGuardians of the Galaxy       1\nPromet...

In [32]:
# Truy xuất giá trị Rank nhỏ hơn 500 trong data frame 
# và chỉ truy show ra 2 cột 'Director','Rating'
movies_df[movies_df['Rank']<500][['Director','Rating']]

Director  Rating
Title                                                
Guardians of the Galaxy            James Gunn     8.1
Prometheus                       Ridley Scott     7.0
Split                      M. Night Shyamalan     7.3
Sing                     Christophe Lourdelet     7.2
Suicide Squad                      David Ayer     6.2
...                                       ...     ...
Under the Skin                Jonathan Glazer     6.3
I Am Legend                  Francis Lawrence     7.2
Men in Black 3               Barry Sonnenfeld     6.8
Super 8                           J.J. Abrams     7.1
Law Abiding Citizen              F. Gary Gray     7.4

[499 rows x 2 columns]

Truy xuất kết hợp nhiều điều kiện

In [33]:
movies_df[(movies_df['Rank']<500) & (movies_df['Rating']>2.5)]

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Under the Skin            495       Drama,Horror,Sci-Fi   
I Am Legend               496       Drama,Horror,Sci-Fi   
Men in Black 3            497   Action,Adventure,Comedy   
Super 8                   498   Mystery,Sci-Fi,Thriller   
Law Abiding Citizen       499      Crime,Drama,Thriller   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Under the Skin           A mysterious young woman seduces lonely men in...   
I Am Legend              Years after a plague kills most of humanity an...   
Men in Black 3           Agent J travels in time to M.I.B.'s early days...   
Super 8                  During the summer of 1979, a group of friends ...   
Law Abiding Citizen      A frustrated man decides to take justice into ...   

                                     Director  Year  Runtime (Minutes)  \
Title                                                                    
Guardians of the Galaxy            James Gunn  2014                121   
Prometheus                       Ridley Scott  2012                124   
Split                      M. Night Shyamalan  2016                117   
Sing                     Christophe Lourdelet  2016                108   
Suicide Squad                      David Ayer  2016                123   
...                                       ...   ...                ...   
Under the Skin                Jonathan Glazer  2013                108   
I Am Legend                  Francis Lawrence  2007                101   
Men in Black 3               Barry Sonnenfeld  2012                106   
Super 8                           J.J. Abrams  2011                112   
Law Abiding Citizen              F. Gary Gray  2009                109   

                         Rating   Votes  Revenue (Millions)  \
Title                                                         
Guardians of the Galaxy     8.1  757074              333.13   
Prometheus                  7.0  485820              126.46   
Split                       7.3  157606              138.12   
Sing                        7.2   60545              270.32   
Suicide Squad               6.2  393727              325.02   
...                         ...     ...                 ...   
Under the Skin              6.3   94707                2.61   
I Am Legend                 7.2  565721              256.39   
Men in Black 3              6.8  278379              179.02   
Super 8                     7.1  298913              126.98   
Law Abiding Citizen         7.4  228339               73.34   

                                                                       new  
Title                                                                       
Guardians of the Galaxy  Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus               Title\nGuardians of the Galaxy       1\nPromet...  
Split                    Title\nGuardians of the Galaxy       1\nPromet...  
Sing                     Title\nGuardians of the Galaxy       1\nPromet...

### 5. Replacing Values, Rename Columns
#### Handling duplicates
This dataset does not have duplicate rows, but it is always important to verify you aren't aggregating duplicate rows.

To demonstrate, let's simply just double up our movies DataFrame by appending it to itself:

In [34]:
temp_df = movies_df.append(movies_df)

temp_df.shape

C:\Users\ADMIN\AppData\Local\Temp/ipykernel_12056/3236548275.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  temp_df = movies_df.append(movies_df)


(2000, 10)

Now we can try dropping duplicates:

In [35]:
temp_df = temp_df.drop_duplicates()

temp_df.shape


(1000, 10)

Now our temp_df will have the transformed data automatically.

Another important argument for drop_duplicates() is keep, which has three possible options:

    first: (default) Drop duplicates except for the first occurrence.
    last: Drop duplicates except for the last occurrence.
    False: Drop all duplicates.

Since we didn't define the keep arugment in the previous example it was defaulted to first. This means that if two rows are the same pandas will drop the second row and keep the first row. Using last has the opposite effect: the first row is dropped.

#### Column cleanup
Many times datasets will have verbose column names with symbols, upper and lowercase words, spaces, and typos. To make selecting data by column name easier we can spend a little time cleaning up their names.

Here's how to print the column names of our dataset:

In [36]:
movies_df.columns

Index(['Rank', 'Genre', 'Description', 'Director', 'Year', 'Runtime (Minutes)',
       'Rating', 'Votes', 'Revenue (Millions)', 'new'],
      dtype='object')

Not only does .columns come in handy if you want to rename columns by allowing for simple copy and paste, it's also useful if you need to understand why you are receiving a Key Error when selecting data by column.

We can use the .rename() method to rename certain or all columns via a dict. We don't want parentheses, so let's rename those:

In [37]:
movies_df.rename(columns={
        'Runtime (Minutes)': 'Runtime', 
        'Revenue (Millions)': 'Revenue_millions'
    }, inplace=True)


movies_df.columns


Index(['Rank', 'Genre', 'Description', 'Director', 'Year', 'Runtime', 'Rating',
       'Votes', 'Revenue_millions', 'new'],
      dtype='object')

Excellent. But what if we want to lowercase all names? Instead of using .rename() we could also set a list of names to the columns like so:

In [38]:
movies_df.columns =  [col.lower() for col in movies_df]

movies_df.columns

Index(['rank', 'genre', 'description', 'director', 'year', 'runtime', 'rating',
       'votes', 'revenue_millions', 'new'],
      dtype='object')

#### How to work with missing values
When exploring data, you’ll most likely encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly you'll see Python's None or NumPy's np.nan, each of which are handled differently in some situations.

There are two options in dealing with nulls:

    Get rid of rows or columns with nulls
    Replace nulls with non-null values, a technique known as imputation

Let's calculate to total number of nulls in each column of our dataset. The first step is to check which cells in our DataFrame are null:

In [39]:
movies_df.isnull()

rank  genre  description  director   year  runtime  \
Title                                                                          
Guardians of the Galaxy  False  False        False     False  False    False   
Prometheus               False  False        False     False  False    False   
Split                    False  False        False     False  False    False   
Sing                     False  False        False     False  False    False   
Suicide Squad            False  False        False     False  False    False   
...                        ...    ...          ...       ...    ...      ...   
Secret in Their Eyes     False  False        False     False  False    False   
Hostel: Part II          False  False        False     False  False    False   
Step Up 2: The Streets   False  False        False     False  False    False   
Search Party             False  False        False     False  False    False   
Nine Lives               False  False        False     False  False    False   

                         rating  votes  revenue_millions    new  
Title                                                            
Guardians of the Galaxy   False  False             False  False  
Prometheus                False  False             False  False  
Split                     False  False             False  False  
Sing                      False  False             False  False  
Suicide Squad             False  False             False  False  
...                         ...    ...               ...    ...  
Secret in Their Eyes      False  False              True  False  
Hostel: Part II           False  False             False  False  
Step Up 2: The Streets    False  False             False  False  
Search Party              False  False              True  False  
Nine Lives                False  False             False  False  

[1000 rows x 10 columns]

Notice isnull() returns a DataFrame where each cell is either True or False depending on that cell's null status.

To count the number of nulls in each column we use an aggregate function for summing:

In [40]:
movies_df.isnull().sum()

rank                  0
genre                 0
description           0
director              0
year                  0
runtime               0
rating                0
votes                 0
revenue_millions    128
new                   0
dtype: int64

We can see now that our data has 128 missing values for revenue_millions and 64 missing values for metascore.

#### Removing null values

Data Scientists and Analysts regularly face the dilemma of dropping or imputing null values, and is a decision that requires intimate knowledge of your data and its context. Overall, removing null data is only suggested if you have a small amount of missing data.

Remove nulls is pretty simple:

In [41]:
movies_df.dropna()

rank                     genre  \
Title                                                      
Guardians of the Galaxy      1   Action,Adventure,Sci-Fi   
Prometheus                   2  Adventure,Mystery,Sci-Fi   
Split                        3           Horror,Thriller   
Sing                         4   Animation,Comedy,Family   
Suicide Squad                5  Action,Adventure,Fantasy   
...                        ...                       ...   
Resident Evil: Afterlife   994   Action,Adventure,Horror   
Project X                  995                    Comedy   
Hostel: Part II            997                    Horror   
Step Up 2: The Streets     998       Drama,Music,Romance   
Nine Lives                1000     Comedy,Family,Fantasy   

                                                                description  \
Title                                                                         
Guardians of the Galaxy   A group of intergalactic criminals are forced ...   
Prometheus                Following clues to the origin of mankind, a te...   
Split                     Three girls are kidnapped by a man with a diag...   
Sing                      In a city of humanoid animals, a hustling thea...   
Suicide Squad             A secret government agency recruits some of th...   
...                                                                     ...   
Resident Evil: Afterlife  While still out to destroy the evil Umbrella C...   
Project X                 3 high school seniors throw a birthday party t...   
Hostel: Part II           Three American college students studying abroa...   
Step Up 2: The Streets    Romantic sparks occur between two dance studen...   
Nine Lives                A stuffy businessman finds himself trapped ins...   

                                      director  year  runtime  rating   votes  \
Title                                                                           
Guardians of the Galaxy             James Gunn  2014      121     8.1  757074   
Prometheus                        Ridley Scott  2012      124     7.0  485820   
Split                       M. Night Shyamalan  2016      117     7.3  157606   
Sing                      Christophe Lourdelet  2016      108     7.2   60545   
Suicide Squad                       David Ayer  2016      123     6.2  393727   
...                                        ...   ...      ...     ...     ...   
Resident Evil: Afterlife    Paul W.S. Anderson  2010       97     5.9  140900   
Project X                      Nima Nourizadeh  2012       88     6.7  164088   
Hostel: Part II                       Eli Roth  2007       94     5.5   73152   
Step Up 2: The Streets              Jon M. Chu  2008       98     6.2   70699   
Nine Lives                    Barry Sonnenfeld  2016       87     5.3   12435   

                          revenue_millions  \
Title                                        
Guardians of the Galaxy             333.13   
Prometheus                          126.46   
Split                               138.12   
Sing                                270.32   
Suicide Squad                       325.02   
...                                    ...   
Resident Evil: Afterlife             60.13   
Project X                            54.72   
Hostel: Part II                      17.54   
Step Up 2: The Streets               58.01   
Nine Lives                           19.64   

                                                                        new  
Title                                                                        
Guardians of the Galaxy   Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus                Title\nGuardians of the Galaxy       1\nPromet...  
Split                     Title\nGuardians of the Galaxy       1\nPromet...  
Sing                      Title\nGuardians of the Galaxy       1\nPromet...  
Suicide Squad             Title\nGuardians of the Galaxy       1\nPromet...  
...               

This operation will delete any row with at least a single null value, but it will return a new DataFrame without altering the original one. You could specify inplace=True in this method as well.

So in the case of our dataset, this operation would remove 128 rows where revenue_millions is null and 64 rows where metascore is null. This obviously seems like a waste since there's perfectly good data in the other columns of those dropped rows. That's why we'll look at imputation next.

Other than just dropping rows, you can also drop columns with null values by setting axis=1:

In [42]:
movies_df.dropna(axis=1)

rank                     genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     director  year  runtime  rating   votes  \
Title                                                                          
Guardians of the Galaxy            James Gunn  2014      121     8.1  757074   
Prometheus                       Ridley Scott  2012      124     7.0  485820   
Split                      M. Night Shyamalan  2016      117     7.3  157606   
Sing                     Christophe Lourdelet  2016      108     7.2   60545   
Suicide Squad                      David Ayer  2016      123     6.2  393727   
...                                       ...   ...      ...     ...     ...   
Secret in Their Eyes                Billy Ray  2015      111     6.2   27585   
Hostel: Part II                      Eli Roth  2007       94     5.5   73152   
Step Up 2: The Streets             Jon M. Chu  2008       98     6.2   70699   
Search Party                   Scot Armstrong  2014       93     5.6    4881   
Nine Lives                   Barry Sonnenfeld  2016       87     5.3   12435   

                                                                       new  
Title                                                                       
Guardians of the Galaxy  Title\nGuardians of the Galaxy       1\nPromet...  
Prometheus               Title\nGuardians of the Galaxy       1\nPromet...  
Split                    Title\nGuardians of the Galaxy       1\nPromet...  
Sing                     Title\nGuardians of the Galaxy       1\nPromet...  
Suicide Squad            Title\nGuardians of the Galaxy       1\nPromet...  
...                                                                    ...  
Secret in Their Eyes     Title\nGuardians of the Galaxy       1\nPromet...  
Hostel: Part II          Title\nGuardians of the Galaxy       1\nPromet...  
Step Up 2: The Streets   Title\nGuardians of the Galaxy       1\nPromet...  
Search Party             Title\nGuardians of the Galaxy       1\nPromet...  
Nine Lives               Title\nGuardians of the Galaxy       1\nPromet...  

[1000 rows x 9 columns]

In our dataset, this operation would drop the revenue_millions and metascore columns
#### Imputation
There may be instances where dropping every row with a null value removes too big a chunk from your dataset, so instead we can impute that null with another value, usually the mean or the median of that column.

Let's look at imputing the missing values in the revenue_millions column. First we'll extract that column into its own variable:

In [43]:
revenue = movies_df['revenue_millions']

In [44]:
revenue.head()

Title
Guardians of the Galaxy    333.13
Prometheus                 126.46
Split                      138.12
Sing                       270.32
Suicide Squad              325.02
Name: revenue_millions, dtype: float64

In [45]:
revenue_mean = revenue.mean()

With the mean, let's fill the nulls using fillna():

In [46]:
revenue.fillna(revenue_mean, inplace=True)

In [47]:
movies_df.isnull().sum()

rank                0
genre               0
description         0
director            0
year                0
runtime             0
rating              0
votes               0
revenue_millions    0
new                 0
dtype: int64

### 6. Lọc dữ liệu trong DataFrame
#### .where()
Việc select các giá trị từ một Series với một Filter (véc tơ boolean) thường trả về một tập con của dữ liệu. Để đảm bảo rằng đầu ra của select có cùng hình dạng với dữ liệu ban đầu, bạn có thể sử dụng phương thức where trong Series và Dataframe. Ví dụ:

In [48]:
s = pd.Series([12,34,56,78],index = [1,2,3,4])

In [49]:
s

1    12
2    34
3    56
4    78
dtype: int64

In [27]:
s>50

1    False
2    False
3     True
4     True
dtype: bool

In [28]:
s[s>50]

3    56
4    78
dtype: int64

In [29]:
s.where(s>50)

1     NaN
2     NaN
3    56.0
4    78.0
dtype: float64

Tuy nhiên, trong trường hợp dùng “where”có một đối số tùy chọn khác để thay thế các giá trị mà điều kiện là False, trong bản sao được trả về.

In [30]:
s.where(s>50,-s)

1   -12
2   -34
3    56
4    78
dtype: int64

#### .query(expr, inplace=False, **kwargs)

Các đối tượng dataframe có một phương thức query() cho phép select dữ liệu thông qua việc sử dụng một biểu thức.

Bạn có thể lấy giá trị của dataframe mà cột b có các giá trị nằm giữa các giá trị của các cột a và c. Ví dụ sau đây sẽ trình bày 2 cách cho cùng một kết quả trả về. Cách một dùng cú pháp giống numy thông thường, cách còn lại dùng theo cú pháp sql.

In [35]:
df = pd.DataFrame(np.random.randint(100, size=(10, 3)), columns=('a b c').split())
df

a   b   c
0  79  68  68
1  61  18  51
2  14  28  17
3  87  46  52
4  16  70  71
5  84  10  62
6  96  57  23
7  86  85  26
8  76  66  54
9  17  65  57

In [36]:
df[(df.a < df.b) & (df.b < df.c)]

a   b   c
4  16  70  71

In [37]:
df.query('a < b and b < c')

a   b   c
4  16  70  71

#### Toán tử ‘in’ và ‘not in’

Query() cũng hỗ trợ sử dụng toán tử so sánh đặc biệt trong Python “in” và “not in”, cung cấp một cú pháp ngắn gọn để gọi phương thức isin() của Series hoặc DataFrame.

In [44]:
df = pd.DataFrame({'a': ('a a b b c c d d e e f f').split(), 'b': ('a a a a b b b b c c c c').split(),'c': np.random.randint(5, size=12), 'd': np.random.randint(9, size=12)})
df

a  b  c  d
0   a  a  1  2
1   a  a  3  7
2   b  a  0  4
3   b  a  0  5
4   c  b  1  5
5   c  b  4  2
6   d  b  1  6
7   d  b  2  5
8   e  c  4  5
9   e  c  3  8
10  f  c  2  1
11  f  c  3  4

In [45]:
df.query('a in b')

a  b  c  d
0  a  a  1  2
1  a  a  3  7
2  b  a  0  4
3  b  a  0  5
4  c  b  1  5
5  c  b  4  2

In [46]:
 df[df.a.isin(df.b)]

a  b  c  d
0  a  a  1  2
1  a  a  3  7
2  b  a  0  4
3  b  a  0  5
4  c  b  1  5
5  c  b  4  2

In [47]:
 df.query('a not in b')

a  b  c  d
6   d  b  1  6
7   d  b  2  5
8   e  c  4  5
9   e  c  3  8
10  f  c  2  1
11  f  c  3  4

In [48]:
df[~df.a.isin(df.b)]

a  b  c  d
6   d  b  1  6
7   d  b  2  5
8   e  c  4  5
9   e  c  3  8
10  f  c  2  1
11  f  c  3  4

Các bộ Filter có thể được kết hợp với nhau qua các toán tử logic như &, |, ~ hay or, and, != … như trong SQL. Cần chú ý rằng mỗi bộ Filter con cần được đặt trong dấu ngoặc ().

Bạn có thể kết hợp “in” hay “not in” với các biểu thức khác cho các truy vấn rất ngắn gọn:

In [49]:
df.query('a in b and c < d')

a  b  c  d
0  a  a  1  2
1  a  a  3  7
2  b  a  0  4
3  b  a  0  5
4  c  b  1  5

In [50]:
df[df.a.isin(df.b) & (df.c < df.d)]

a  b  c  d
0  a  a  1  2
1  a  a  3  7
2  b  a  0  4
3  b  a  0  5
4  c  b  1  5

#### .lockup(row_lables, col_labels)

In [55]:
dflookup = pd.DataFrame(np.random.rand(5,4), columns = ['A','B','C','D'])

In [56]:
dflookup

A         B         C         D
0  0.801301  0.895562  0.755226  0.669032
1  0.825270  0.062641  0.780621  0.067138
2  0.132254  0.727103  0.251096  0.547512
3  0.013428  0.071927  0.137375  0.341412
4  0.576656  0.207708  0.923489  0.257744

In [62]:
dflookup.lookup([0, 2, 4], ['B','C','A'])

C:\Users\ADMIN\AppData\Local\Temp/ipykernel_18912/541526541.py:1: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.
  dflookup.lookup([0, 2, 4], ['B','C','A'])


array([0.89556177, 0.25109612, 0.57665558])

### 7. Xác định các tham số thống kê: Sum, Cumsum, Min, Max, Mean, Median, Std
Using describe() on an entire DataFrame we can get a summary of the distribution of continuous variables:

In [50]:
movies_df.describe()

rank         year      runtime       rating         votes  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000   113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962    18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000    66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000   100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000   111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000   123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000   191.000000     9.000000  1.791916e+06   

       revenue_millions  
count       1000.000000  
mean          82.956376  
std           96.412043  
min            0.000000  
25%           17.442500  
50%           60.375000  
75%           99.177500  
max          936.630000

.describe() can also be used on a categorical variable to get the count of rows, unique count of categories, top category, and freq of top category:

In [51]:
movies_df['genre'].describe()

count                        1000
unique                        207
top       Action,Adventure,Sci-Fi
freq                           50
Name: genre, dtype: object

This tells us that the genre column has 207 unique values, the top value is Action/Adventure/Sci-Fi, which shows up 50 times (freq).

.value_counts() can tell us the frequency of all values in a column:

In [52]:
movies_df['genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Action,Adventure,Fantasy      27
Comedy,Drama                  27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: genre, dtype: int64

#### Cumsum
Return cumulative sum over a DataFrame or Series axis.

Returns a DataFrame or Series of the same size containing the cumulative sum.

In [67]:
s = pd.Series([2, np.nan, 5, -1, 0])

s

0    2.0
1    NaN
2    5.0
3   -1.0
4    0.0
dtype: float64

In [68]:
s.cumsum()

0    2.0
1    NaN
2    7.0
3    6.0
4    6.0
dtype: float64

To include NA values in the operation, use skipna=False

In [69]:
s.cumsum(skipna=False)

0    2.0
1    NaN
2    NaN
3    NaN
4    NaN
dtype: float64

In [70]:
df = pd.DataFrame([[2.0, 1.0],

                   [3.0, np.nan],

                   [1.0, 0.0]],

                   columns=list('AB'))

df

A    B
0  2.0  1.0
1  3.0  NaN
2  1.0  0.0

By default, iterates over rows and finds the sum in each column. This is equivalent to axis=None or axis='index'.

In [71]:
df.cumsum()

A    B
0  2.0  1.0
1  5.0  NaN
2  6.0  1.0

To iterate over columns and find the sum in each row, use axis=1

In [72]:
df.cumsum(axis=1)

A    B
0  2.0  3.0
1  3.0  NaN
2  1.0  1.0

### 8. Giá trị duy nhất (Unique)
Return unique values based on a hash table.

Uniques are returned in order of appearance. This does NOT sort.

Significantly faster than numpy.unique for long enough sequences. Includes NA values.

In [63]:
pd.unique(pd.Series([2, 1, 3, 3]))

array([2, 1, 3], dtype=int64)

In [66]:
del list
pd.unique(
    pd.Series(
        pd.Categorical(list("baabc"), categories=list("abc"), ordered=True)
    )
)

['b', 'a', 'c']
Categories (3, object): ['a' < 'b' < 'c']

In [74]:
pd.unique([("a", "b"), ("b", "a"), ("a", "c"), ("b", "a")])

array([('a', 'b'), ('b', 'a'), ('a', 'c')], dtype=object)

In [75]:
pd.unique(
    pd.Series(
        [
            pd.Timestamp("20160101", tz="US/Eastern"),
            pd.Timestamp("20160101", tz="US/Eastern"),
        ]
    )
)

<DatetimeArray>
['2016-01-01 00:00:00-05:00']
Length: 1, dtype: datetime64[ns, US/Eastern]